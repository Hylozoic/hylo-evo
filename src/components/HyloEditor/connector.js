import { connect } from 'react-redux'
import { findMentions, clearMentions, getMentionResults } from './store'

export function mapStateToProps (state, props) {
  return {
    mentionResults: getMentionResults(state, props)
  }
}

export const mapDispatchToProps = {
  findMentions,
  clearMentions
}

export default connect(mapStateToProps, mapDispatchToProps)
