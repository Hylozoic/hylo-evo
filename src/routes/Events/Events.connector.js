import { get, isEmpty } from 'lodash/fp'
import { push } from 'connected-react-router'
import { connect } from 'react-redux'
import fetchPosts from 'store/actions/fetchPosts'
import {
  FETCH_POSTS, FETCH_FOR_CURRENT_USER
} from 'store/constants'
import presentPost from 'store/presenters/presentPost'
import getGroupForCurrentRoute from 'store/selectors/getGroupForCurrentRoute'
import getRouteParam from 'store/selectors/getRouteParam'
import getMe from 'store/selectors/getMe'
import getMyMemberships from 'store/selectors/getMyMemberships'
import { getHasMorePosts, getPosts } from 'store/selectors/getPosts'
import { createPostUrl } from 'util/navigation'

import { updateTimeframe } from './Events.store'

export function mapStateToProps (state, props) {
  let group

  const routeParams = get('match.params', props)
  const location = get('location', props)
  const currentUser = getMe(state)
  const currentUserHasMemberships = !isEmpty(getMyMemberships(state))
  const groupSlug = getRouteParam('groupSlug', state, props)

  if (groupSlug) {
    group = getGroupForCurrentRoute(state, props)
  }

  const timeframe = state.Events.timeframe

  const fetchPostsParam = {
    afterTime: timeframe === 'future' ? new Date().toISOString() : null,
    beforeTime: timeframe === 'past' ? new Date().toISOString() : null,
    context: routeParams.context,
    filter: 'event',
    order: timeframe === 'future' ? 'asc' : 'desc',
    slug: groupSlug,
    sortBy: 'start_time'
  }

  // NOTE: In effort to better seperate the query caching from component details
  //       it's better (and necessary) in this case to send the fetch param then
  //       the raw props of the component.
  const posts = getPosts(state, fetchPostsParam).map(p => presentPost(p, get('id', group)))
  const hasMore = getHasMorePosts(state, fetchPostsParam)

  return {
    routeParams,
    location,
    currentUser,
    currentUserHasMemberships,
    fetchPostsParam,
    groupSlug,
    group,
    hasMore,
    membershipsPending: state.pending[FETCH_FOR_CURRENT_USER],
    posts,
    pending: state.pending[FETCH_POSTS],
    timeframe
  }
}

export function mapDispatchToProps (dispatch, props) {
  const routeParams = get('match.params', props)

  return {
    fetchEvents: (params) => (offset) => dispatch(fetchPosts({ offset, ...params })),
    updateTimeframe: timeframe => dispatch(updateTimeframe(timeframe)),
    newPost: () => dispatch(push(createPostUrl(routeParams, { newPostType: 'event' })))
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { fetchPostsParam } = stateProps

  return {
    ...ownProps,
    ...stateProps,
    ...dispatchProps,
    fetchEvents: dispatchProps.fetchEvents(fetchPostsParam)
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
