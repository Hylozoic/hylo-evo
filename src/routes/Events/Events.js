import cx from 'classnames'
import { get } from 'lodash/fp'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import Dropdown from 'components/Dropdown'
import FeedBanner from 'components/FeedBanner'
import Icon from 'components/Icon'
import Loading from 'components/Loading'
import NoPosts from 'components/NoPosts'
import PostCard from 'components/PostCard'
import ScrollListener from 'components/ScrollListener'
import { CENTER_COLUMN_ID } from 'util/scrolling'
import s from './Events.scss'
import viewStyles from '../../components/StreamViewControls/StreamViewControls.scss'

const timeframeOptions = [
  { id: 'future', label: 'Upcoming Events' },
  { id: 'past', label: 'Past Events' }
]

export default class Events extends Component {
  static propTypes = {
    newPost: PropTypes.func,
    routeParams: PropTypes.object,
    querystringParams: PropTypes.object
  }

  static defaultProps = {
    routeParams: {},
    querystringParams: {}
  }

  componentDidMount () {
    this.props.fetchEvents()
  }

  componentDidUpdate (prevProps) {
    const { routeParams, fetchEvents } = this.props
    const { context, groupSlug } = routeParams

    if ((context && get('routeParams.context', prevProps) !== context) ||
        (groupSlug && get('routeParams.groupSlug', prevProps) !== groupSlug) ||
        prevProps.timeframe !== this.props.timeframe) {
      fetchEvents()
    }
  }

  fetchMoreEvents = () => {
    const { pending, posts, hasMore, fetchEvents } = this.props
    if (pending || posts.length === 0 || !hasMore) return
    fetchEvents(posts.length)
  }

  render () {
    const {
      collapsedState,
      routeParams, querystringParams, group, currentUser,
      newPost, currentUserHasMemberships,
      membershipsPending, pending, posts,
      timeframe, updateTimeframe
    } = this.props
    const { context } = routeParams

    if (!currentUser) return <Loading />
    if (membershipsPending) return <Loading />

    return <div>
      <FeedBanner
        group={group}
        currentUser={currentUser}
        type='event'
        context={context}
        newPost={newPost}
        routeParams={routeParams}
        querystringParams={querystringParams}
        currentUserHasMemberships={currentUserHasMemberships} />

      <div styleName='viewStyles.stream-view-ctrls'>
        <Dropdown className={viewStyles.dropdown}
          toggleChildren={<span styleName='viewStyles.dropdown-label'>
            <Icon name='ArrowDown' />
            {timeframeOptions.find(o => o.id === timeframe).label}
          </span>}
          items={timeframeOptions.map(({ id, label }) => ({
            label,
            onClick: () => updateTimeframe(id)
          }))}
        />
      </div>

      {pending && <Loading />}

      <div styleName={cx('s.events-stream', { [s.collapsedState]: collapsedState })}>
        {!pending && posts.length === 0 ? <NoPosts message={`No ${timeframe === 'future' ? 'upcoming' : 'past'} events`} /> : ''}

        {posts.map(post => {
          const expanded = post.id === routeParams.postId
          return <PostCard
            routeParams={routeParams}
            post={post}
            styleName={cx('s.event-card', { 's.expanded': expanded })}
            expanded={expanded}
            key={post.id} />
        })}
      </div>
      <ScrollListener onBottom={this.fetchMoreEvents} elementId={CENTER_COLUMN_ID} />

    </div>
  }
}
