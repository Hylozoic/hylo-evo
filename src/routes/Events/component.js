import React from 'react'
import { Link } from 'react-router-dom'
import './component.scss'

const { string, bool } = React.PropTypes

export default function Events () {
  return <div styleName='event-list'>
    <div><Link to='/events/1'>Event 1</Link></div>
    <div><Link to='/events/2'>Event 2</Link></div>
    <div><Link to='/events/3'>Event 3</Link></div>
    <div><Link to='/events/4'>Event 4</Link></div>
  </div>
}
Events.propTypes = {
  label: string,
  color: string,
  active: bool,
  className: string
}
