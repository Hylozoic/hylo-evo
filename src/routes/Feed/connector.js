import { connect } from 'react-redux'
import { SAMPLE_FEED_ITEMS } from './sampleData'

function mapStateToProps (state) {
  return {
    feedItems: SAMPLE_FEED_ITEMS
  }
}

function mapDispatchToProps (dispatch, props) {
  return {}
}

export default connect(mapStateToProps, mapDispatchToProps)
