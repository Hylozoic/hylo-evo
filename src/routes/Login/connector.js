import { connect } from 'react-redux'
import { checkLogin, login } from './actions'
export default connect(null, {checkLogin, login})
