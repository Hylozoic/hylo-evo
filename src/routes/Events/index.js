import component from './Events'
import connector from './Events.connector'

export default connector(component)
