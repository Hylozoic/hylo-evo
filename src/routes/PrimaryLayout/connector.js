import { connect } from 'react-redux'
import { SAMPLE_COMMUNITY } from 'routes/Feed/sampleData'
import { fetchCurrentUser } from './actions'

function mapStateToProps ({currentUser}) {
  return {
    community: SAMPLE_COMMUNITY,
    currentUser
  }
}

export default connect(mapStateToProps, {fetchCurrentUser})
