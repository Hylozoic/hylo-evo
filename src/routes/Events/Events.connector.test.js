/* eslint no-unused-vars:'off' */
import orm from 'store/models'
import { mapStateToProps, mapDispatchToProps } from './Events.connector'

describe('mapStateToProps', () => {
  let session, state

  const matchProps = {
    params: {
      groupSlug: 'foo'
    }
  }

  beforeEach(() => {
    session = orm.session(orm.getEmptyState())
    session.Group.create({ id: '1', slug: 'foo', postCount: 10 })
    state = {
      pending: {},
      orm: session.state,
      Events: { timeframe: 'past' }
    }
  })

  it('sets group from the router props', () => {
    const props = {
      match: matchProps
    }
    expect(mapStateToProps(state, props)).toMatchObject({
      group: expect.objectContaining({ id: '1', slug: 'foo' })
    })
  })
})

describe('mapDispatchToProps', () => {
  let state

  beforeEach(() => {
    state = {
      pending: {},
      Events: { timeframe: 'past' }
    }
  })

  it('correctly sets up updateTimeframe and fetchEvents', () => {
    const dispatch = jest.fn(x => x)
    const props = {
      location: {
        pathname: '/groups/foo'
      },
      match: {
        params: { groupSlug: 'foo' }
      }
    }

    const fetchPostsParam = {
      afterTime: new Date().toISOString(),
      context: 'group',
      filter: 'event',
      order: 'asc',
      slug: 'foo',
      sortBy: 'start_time'
    }

    const dispatchProps = mapDispatchToProps(dispatch, props)
    expect(dispatchProps.updateTimeframe('future')).toMatchSnapshot()
    expect(dispatchProps.fetchEvents(fetchPostsParam)).toMatchSnapshot()
    expect(dispatchProps.newPost(fetchPostsParam)).toMatchSnapshot()
  })
})
