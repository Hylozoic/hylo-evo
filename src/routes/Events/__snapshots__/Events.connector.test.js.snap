// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps correctly sets up updateTimeframe and fetchEvents 1`] = `
Object {
  "payload": "future",
  "type": "Events/UPDATE_TIMEFRAME",
}
`;

exports[`mapDispatchToProps correctly sets up updateTimeframe and fetchEvents 2`] = `[Function]`;

exports[`mapDispatchToProps correctly sets up updateTimeframe and fetchEvents 3`] = `
Object {
  "payload": Object {
    "args": Array [
      "/groups/foo/create/post?newPostType=event",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;
